nextcloud: update a few defaults

Ma27 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Please review by commit.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
nixos/nextcloud: drop default for `dbtype` setting

Closes #355606

To quote the system requirements[1]:

> Database: SQLite 3.16+ (only recommended for testing and minimal-instances)

As discussed[2], this would already fail at eval-time for everyone and
makes all sqlite users aware that this is probably not the best choice.
Doing that regardless, is now an explicit decision.

[1] https://docs.nextcloud.com/server/30/admin_manual/installation/system_requirements.html
[2] https://github.com/NixOS/nixpkgs/pull/369242#discussion_r1899706512

Ma27 committed:
nixos/nextcloud: increase pool settings

Also link to upstream docs about PHP-FPM tuning[1].
These seem a little more reasonable performance-wise, but are still
conservative enough for small setups.

[1] https://docs.nextcloud.com/server/30/admin_manual/installation/server_tuning.html#tune-php-fpm

Shawn8901 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -416,7 +420,8 @@ in {
       dbtype = mkOption {
         type = types.enum [ "sqlite" "pgsql" "mysql" ];
-        default = "sqlite";
+        default = if versionAtLeast "25.05" config.system.stateVersion then "pgsql" else "sqlite";
```
does it make sense to enable `createLocally` too when defaulting for postgres?

Ma27 replied:
This would silently activate a postgresql for everyone who has nextcloud already configured, but with a remote database.
So I'd prefer to keep this opt-in.

Shawn8901 replied:
i would tie it to the stateVersion also, but i am also okay in keeping it as it is.

Ma27 replied:
Now that the defaults are dropped anyways, I think this can be resolved.

Atemu commented on nixos/tests/nextcloud/basic.nix:
```diff
@@ -59,6 +59,7 @@ runTest (
 
           services.nextcloud = {
             enable = true;
+            config.dbtype = "sqlite";
```
Shouldn't this ideally also test the default?

Ma27 replied:
Hmm, fair.
I'm inclined to go for
* `basic` testing all defaults (i.e. postgresql)
* `with-sqlite-and-redis` to have Redis and sqlite coverage (it's good to have coverage for each DB and caching approach, but I don't recall a single issue in several years of maintaining this that we would've caught by testing all combinations of that).

Atemu replied:
That sounds good to me.

Ma27 replied:
Hmm this won't work out as planned given that notify push is also part of that test and this doesn't seem to work with sqlite (though it doesn't complain when sqlite is configured and just fails at runtime cc @SuperSandro2000)

Ma27 replied:
Fixed.

Atemu commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -416,7 +420,8 @@ in {
       dbtype = mkOption {
         type = types.enum [ "sqlite" "pgsql" "mysql" ];
-        default = "sqlite";
+        default = if versionAtLeast "25.05" config.system.stateVersion then "pgsql" else "sqlite";
```
Perhaps we should also put out an eval warning that sqlite is being used despite being unfit for production.

Though we can't really recommend existing users to do anything about it given that database migration is *still* broken upstream :/

The intention would be to prevent users like me who have an old stateVersion on their existing machine from unknowingly using sqlite when that's something you really don't want with (currently) not even a way out.

We could perhaps even leave the default value unset, requiring users to explicitly pick their poison.  
Existing users too would need to explicitly choose their current database which would theoretically be a "breaking" change but hardly a bad one and it's arguably better to be a little more explicit here anyways.

Ma27 replied:
> Perhaps we should also put out an eval warning that sqlite is being used despite being unfit for production.

Non-actioanble eval warnings are something I'd like to avoid, that only results in people ignoring warnings. There's this problem-infrastructure that we can probably use (I don't know if it's nixpkgs-only or also useful for nixos) if it gets merged at some point.

> The intention would be to prevent users like me who have an old stateVersion on their existing machine from unknowingly using sqlite when that's something you really don't want with (currently) not even a way out.

The release notes explicitly mention this. So I'd argue the next time people read these release-notes, they'll be aware (this is of course based on my _expectation_ that people read release-notes).

> Existing users too would need to explicitly choose their current database which would theoretically be a "breaking" change but hardly a bad

I mean that would only break people who haven't set their dbtype already, i.e. sqlite users only though. So perhaps that's the way to go.

Atemu replied:
Yeah I'm most partial towards the last idea.

Ma27 committed:
nixos/nextcloud: don't use pg14 for tests

This currently fails to build, so we just update it now to make sure all
tests are green with the current changes.

Ma27 commented:
@Atemu are you fine with the current state? I'm inclined to submit this tomorrow otherwise.

Atemu reviewed (approved):
I can't speak for the other changes but the diff w.r.t. DB default LGTM.

martinetd commented:
Hi, it's been a while but I hit this as upgrading from 24.12 to 25.05 -- for now it's easy enough to explicitly set sqlite, but I'd be happy to use postgres instead as there's already one running...

How would one go about migrating the db? If nextcloud has a db-agnostic export/import feature perhaps starts as sqlite, dump it all, restart as postgres, import?
If there's any sane way of doing it I think that it should be added to release notes (possibly just as a link to upstream's export feature if the above exists)

Thanks

SuperSandro2000 commented:
Maybe the nextcloud docs can help you? https://docs.nextcloud.com/server/latest/admin_manual/configuration_database/db_conversion.html

martinetd commented:
Done in #422433
